### PR TITLE
client: fix deadlock related to async pagecache invalidation

### PR DESCRIPTION
**Pull request by ukernel**
When mds reovkes Fc capability, client code first invalidates kernel
pagecache, then calls check_caps() to release Fc capability to MDS.
invalidating kernel pagecache can block if there is writer holding
i_mutex. Deadlock happens if the writer needs to wait Fw capability.
(MDS can't finish revoking Fc capability because client doesn't call
 check_caps(). So mds can't issue Fw to the client)

The fix is call check_caps() after invalidating the objectcacher.

Fixes: #13800
Signed-off-by: Yan, Zheng <zyan@redhat.com>

### CODE REVIEW

File: src/client/Client.cc [hunk 1]
```diff
@@ -4689,7 +4685,7 @@ void Client::handle_cap_grant(MetaSession *session, Inode *in, Cap *cap, MClient
         && !_flush(in, new C_Client_FlushComplete(this, in))) {
       // waitin' for flush
     } else if ((old_caps & ~new_caps) & CEPH_CAP_FILE_CACHE) {
-      _release(in);
+      check = _release(in);
```
Comment (jcsp) [inline src/client/Client.cc]: Shouldn't this be "check |= _release(in)" to avoid overriding if check was already true?

**Comment by jcsp** [comment]
Other than the one comment this looks good to me: it needs an FS suite run before merging though.

**Comment by gregsfortytwo** [comment]
@ukernel, pelase describe how you tested this PR.

**Comment by ukernel** [comment]
On Wed, Nov 25, 2015 at 12:02 AM, Gregory Farnum <notifications@github.com>
wrote:

> @ukernel <https://github.com/ukernel>, pelase describe how you tested
> this PR.
>
I don't have test case for this corner case. This bug is reported by an
user, it seems his problem get solved.

https://www.mail-archive.com/ceph-users@lists.ceph.com/msg24747.html

**Comment by gregsfortytwo** [comment]
I'd like if we could get some good tests for this but I think they'd have to wait until we have good multi-client ones for checking capability races. :/ It didn't add any breakage in the tests at http://pulpito.ceph.com/ubuntu-2015-11-24_12:59:17-fs-greg-fs-testing---basic-multi/ so I merged this manually to resolve a conflict.
